app/data/tasks/deduplicate_records.py

```diff
@@ -0,0 +1,187 @@
+        job.status = DedupeJob.Status.SUCCESS
+
+    def on_failure(self, exc, task_id, args, kwargs, einfo):
+        """ On failure, delete duplicates found in the unsuccessful run"""
+        job = DedupeJob.objects.get(celery_task=task_id)
+        job.status = DedupeJob.Status.ERROR
+        RecordDuplicate.objects.filter(job=job).delete()
+
+
+def calculate_similarity_score(record1, record2, time_allowance, distance_allowance):
+    """
+    Set a score representing the likelyhood the two records are duplicated, based
+    on the criteria.  score is a float between 0 and 1
+
+    normalized dimensions are considered equal (1 day equally as important as 100m)
+    """
+    tdelta = abs(record1.occurred_from - record2.occurred_from)
+    tscore = 0
+    tscore = 1.0 - (tdelta.total_seconds() / time_allowance.total_seconds())
+    ddelta = record1.geom.distance(record2.geom)
+    dscore = 0
+    dscore = 1.0 - (ddelta / distance_allowance)
+    score = (tscore + dscore) / 2
+    return score
+
+
+def find_duplicates_for_record(uuid, time_allowance, distance_allowance):
+    """ Given a record, find all possible duplicates for it
+    """
+    record = Record.objects.get(uuid=uuid)
+    return Record.objects.filter(
+        schema__record_type=record.schema.record_type,
+        geom__dwithin=(record.geom, distance_allowance),
+        occurred_from__range=(
+            record.occurred_from - time_allowance, record.occurred_from + time_allowance
+        )
+    ).exclude(
+        uuid=record.uuid,
+        uuid__in=RecordDuplicate.objects.filter(
+            Q(record__uuid=record.uuid) | Q(
+                duplicate_record__uuid=record.uuid
+            )
+        )
+    )
+
+
+def get_time_extent(time_allowance):
+    """ Get the time range to dedupe
+    Calculate from X hr before the last dedupe run to the current task run's
+    start time, where X is the heuristic's time allowance
+
+    Returns:
+        dictionary containing the start and end times for the next dedupe task
+        {starttime: timestamp, endtime: timestamp}
+    """
+    # TODO: get config var determining matching time - use value of 1 hr for now
+    last_job = None
+    if DedupeJob.objects.all().count() > 0:
+        last_job = DedupeJob.objects.latest()
+    start_time = None
+    if last_job:
+        start_time = last_job.datetime
+    else:
+        start_time = Record.objects.earliest('created').created
+    end_time = datetime.datetime.utcnow().replace(tzinfo=timezone('UTC'))
+    return {'start_time': start_time, 'end_time': end_time}
+
+
+def get_dedupe_set(extent):
+    """ Get the set of records to dedupe
+    return a set of uuids to check for duplicates, and the queryset used to
+    generate it
+    """
+    queryset = Record.objects.filter(
+        created__range=(
+            extent['start_time'], extent['end_time']
+        )
+    ).exclude(uuid__in=get_dedupe_ids())
+    ids = queryset.values_list('uuid', flat=True)
```

Comment:
I still have a slight preference for `'pk'` here since it's a stand-in for "whatever key is the primary key", so it'll work with any Django model, not just those that have a primary key called `uuid`. This works for this model though so it's not a major issue.